### ISSUE TITLE
`XMLHttpRequest` failing to construct `Response` when body status is `null`

### ISSUE THREAD

**Issue by disco-panda** [mswjs/interceptors#379]
<img width="1021" alt="Screenshot 2023-05-25 at 12 00 56 PM" src="https://github.com/mswjs/interceptors/assets/66640959/980fe1df-f140-459c-ae3e-b834df6b2512">

<img width="1024" alt="Screenshot 2023-05-25 at 12 05 54 PM" src="https://github.com/mswjs/interceptors/assets/66640959/84d2d7b7-3dbc-4226-a687-81a18668fe57">

<img width="1124" alt="Screenshot 2023-05-25 at 12 13 10 PM" src="https://github.com/mswjs/interceptors/assets/66640959/eda965a3-799a-4193-ae47-e34b12eaa3b4">

When using the `XMLHttpRequestInterceptor` these errors are thrown, crashing the webpage. 

It does not appear when using other interceptors.

This is the code I am using - 

```
import { BatchInterceptor } from '@mswjs/interceptors'
import { FetchInterceptor } from '@mswjs/interceptors/fetch'
import { XMLHttpRequestInterceptor } from '@mswjs/interceptors/XMLHttpRequest'


const interceptor = new BatchInterceptor({
    name: 'my-interceptor',
    interceptors: [new FetchInterceptor(), new XMLHttpRequestInterceptor()],
})

interceptor.apply()

interceptor.on('request', async (e: any, request: any) => {
    console.log("request", await e.clone().json())
})

interceptor.on('response', async (res: any, request: any) => {
    console.log("response", await res.clone().json())
})
```

**Comment by disco-panda** [comment]
Appears similar to #378

**Comment by kettanaito** [comment]
Hi, @disco-panda. Thanks for reporting this. 

So there seems to be two issues going on:

1. The request gets terminated due to the forbidden request header (likely).
2. While the request is terminated, the library still attempts to construct a `Response` with incorrect input data. 

I would love to put this use case into an integration test. This should be fairly easy to emulate with our testing setup. 

The actual bug here is that we should never construct the `Response` instance if the request has been terminated.

**Comment by disco-panda** [comment]
Thank you so much for the fast response and great explanation of what is going on!

**Comment by disco-panda** [comment]
Funnily enough, when I moved to using Vite to build my code instead of webpack the issue went away... hmm.....

**Comment by avivasyuta** [comment]
Hello.

This issue is completely unrelated to passing headers from #378.
The essence of the problem is that the interceptor is trying to construct a `new Response()` with a body for responses that, according [to the specification](https://fetch.spec.whatwg.org/#statuses), cannot have a body. 

[Here](https://github.com/mswjs/interceptors/blob/main/src/interceptors/XMLHttpRequest/utils/createResponse.ts#L9) `body` is equal empty string `""`, but for http statuses `101, 103, 204, 205, 304` it must be `null`.

**Comment by kettanaito** [comment]
@avivasyuta, looks like we need to do a bit of response body transformations since XHR responses don't directly translate to Fetch API responses. Please, would you have a chance to open a pull request with a failing test for this? We can collaborate on it and ship the fix in a timely manner. I would be thankful!

**Comment by avivasyuta** [comment]
Yes, I'm just trying to figure out how to write an test for this scenario. I'll open a PR when it is ready.

**Comment by kettanaito** [comment]
I believe we need to model a two scenarios (translate to 2 tests):

1. An XHR interceptor that has no `request` listener (no mocking) and instead requests an actual HTTP server that responds with one of the responses you mentioned that mustn't have bodies, like a `204 No Content` for a `GET` request. We then add a `response` listener on the test and assert that it returns a valid response representation.
2. An XHR interceptor that has a `request` listener that returns the same response but a mocked one. This is just to make sure nothing goes awry between the consumer and the actual response arriving at the XHR (since the validity of the constructed response is now on the consumer's shoulders we can still, potentially, have a bug in the interceptor that would not handle that response correctly). In this test, we'd instead assert the `xhr.response` returned on the request instance. 

You can take inspiration from the existing tests under `test/modules/XMLHttpRequest`, which are plenty and feature both of the above scenarios.

**Comment by avivasyuta** [comment]
The trick is that the original `XMLHttpRequest` will have `xhr.responseText === ''` even for responses with 204 HTTP status code. Also if we create new response with `null` body than it text variant will equal empty string ``.

```javascript
const resp = new Response(null)
await resp.text() // ''
```

So I can't figure out how to track the changes in the test.
I made a [simple test](https://github.com/mswjs/interceptors/pull/411) that always passes. But in the console we can see that an error occurs.

<img width="1144" alt="Screenshot 2023-09-01 at 21 53 46" src="https://github.com/mswjs/interceptors/assets/15625379/0168da61-8976-4b60-b5bc-4ec8d892a43c">

**Comment by kettanaito** [comment]
I think that's the expected behavior as we channel runtime exceptions from Playwright but don't treat them as test failures to allow us to test those exceptions as intended. 

Looks good! Please open a pull request with that if you haven't already. 

> The trick is that the original XMLHttpRequest will have xhr.response === '' even for responses with 204 HTTP status codes.

I see. Our goal is to manually check the response status code and prevent anything but `null` as the response body in those special cases. 

> Also if we create new response with null body than it text variant will equal empty string ``.

This is also correct. The proper way to get `null` response body back is by `response.body`, which will be a `ReadableStream` if the response has a body to stream, or `null` if the response body was set to `null` when constructing the response.

**Comment by avivasyuta** [comment]
It seems to me that I've realised how to fix and test this issue.
I've made a [PR](https://github.com/mswjs/interceptors/pull/411/files#diff-e56d16939dcd9b168b7992063618f8de000c5e153cb5d883a262e2aa2913c988).

Please see if this is a good approach when you have time so I can proceed.

**Comment by kettanaito** [comment]
## Released: v0.24.1 🎉

This has been released in v0.24.1!

- 📄 [**Release notes**](https://github.com/mswjs/interceptors/releases/tag/v0.24.1)
- 📦 [npm package](https://www.npmjs.com/package/@mswjs/interceptors/v/0.24.1)

Make sure to always update to the latest version (`npm i @mswjs/interceptors@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.
